### PR TITLE
[e2e][parallel] e2e + fix for Class Not Found on Parallel

### LINKED ISSUE #6903: [Parallel]: Class was not found while trying to analyse it
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.12.10            |

## Minimal PHP Code Causing Issue

When enable parallel option, it can got error:

```bash
 [ERROR] Could not process "tests/system/Files/FileCollectionTest.php" file, due to:                                    
         "System error: "Class FileCollectionTest was not found while trying to analyse it - discovering symbols is     
         probably not configured properly."                                                                             
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 45     
```

Step to reproduce:

1. Clone repo

```bash
git clone git@github.com:codeigniter4/CodeIgniter4.git
```

2. Run composer install

```bash
composer install
```

3. Open `rector.php`, add option parall:

```php
$parameters->set(Option::PARALLEL, true);
```

4. Run against `tests` directory:

```bash
vendor/bin/rector process tests
```

## Expected Behaviour

It should show `[OK] Rector is done!` without diff and error:

```bash
vendor/bin/rector process tests 
 257/257 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] Rector is done!                                                                                    
```

### PR DESCRIPTION
**Pull request by samsonasik**
@TomasVotruba here is the e2e tests for Class Not Found on Parallel for issue https://github.com/rectorphp/rector/issues/6903 with run the following command:

```bash
cd e2e/parallel-reflection-resolver
php ../e2eTestRunner.php
```

**With Parallel enabled in `e2e/parallel-reflection-resolver/rector.php`:**, it got Error, ref https://github.com/rectorphp/rector-src/runs/4682788823?check_suite_focus=true#step:6:29

```bash
EXPECTED:
string(300) "1 file with changes
===================

1) src/SomeClass.php:3

    ---------- begin diff ----------
@@ @@

 final class SomeClass
 {
-    private $foo;
 }
    ----------- end diff -----------

Applied rules:
 * RemoveUnusedPrivatePropertyRector


 [OK] 1 file would have changed (dry-run) by Rector"

ACTUAL:
string(348) "[ERROR] Could not process "src/SomeClass.php" file, due to:
         "System error: "Class SomeClass was not found while trying to analyse it - discovering symbols is probably not
         configured properly."
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On
         line: 45"
```

**Without Parallel enabled in `e2e/parallel-reflection-resolver/rector.php`:**, it got "end-to-end test successfully completed".

```diff
 php ../e2eTestRunner.php
[OK] end-to-end test successfully completed.%       
 ```

The rector rule applied is currently:

```php
Rector\DeadCode\Rector\Property\RemoveUnusedPrivatePropertyRector
```

The diff expected is at https://github.com/rectorphp/rector-src/pull/1613/commits/56e4160f94eb130a1301d5da44e887ef097ee12d for test.

Fixes https://github.com/rectorphp/rector/issues/6903

### CODE REVIEW

**Comment by samsonasik** [comment]
It seems because on `runParallel()` on 

https://github.com/rectorphp/rector-src/blob/9c6a6b04228e3af56893d9c995459b37b879e802/src/Application/ApplicationFileProcessor.php#L81-L82

the files information is not yet stored, compared with the else for **non-parallel**:

https://github.com/rectorphp/rector-src/blob/9c6a6b04228e3af56893d9c995459b37b879e802/src/Application/ApplicationFileProcessor.php#L83-L95

File: src/Provider/ReflectionProvider.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public function getClass(string $className): ClassReflection
+    {
+        if (! $this->reflectionProvider->hasClass($className)) {
+            dump('here');
+        }
+
+        return $this->reflectionProvider->getClass($className);
+    }
```
Comment (samsonasik) [inline src/Provider/ReflectionProvider.php]: @TomasVotruba I am trying to override PHPStan's `ReflectionProvider::getClass()` method, any way to create `ClassReflection` instance manually based on the file/nodes?

Comment (TomasVotruba) [reply src/Provider/ReflectionProvider.php]: I never tried to fake my own `ClassReflection` of PHPStan, so I'm not sure how to do that.

Comment (samsonasik) [reply src/Provider/ReflectionProvider.php]: Fixed 🎉 by re-run with fallback to non-parallel for current file pointed from error when got error Class not found https://github.com/rectorphp/rector-src/pull/1613/commits/454acbcbe93b647a4996776267721da2e6407b5c

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by TomasVotruba** [comment]
Hm, it seems the files are not passed to PHPStan properly in parallel. I'll have to look into this

**Comment by samsonasik** [comment]
@TomasVotruba thank you, here is the steps to reproduce locally:

```bash
# needed to register autoload E2e\\ namespace for Class that autoloadable
composer install 

# go to e2e/parallel-reflection-resolver and run e2e runner
cd e2e/parallel-reflection-resolver
php ../e2eTestRunner.php
```

This e2e runner will assert compare with `e2e/parallel-reflection-resolver/expected-output.diff`

For note: if this solution is the only way to go, we may need to override `ReflectionProvider::hasClass()` in separate PR as this PR is cover only direct `ReflectionProvider::getClass()` for a certainly passed `Class_` instance.

**Comment by TomasVotruba** [comment]
I'm looking into this...

**Comment by TomasVotruba** [comment]
Could you rebase on `main`?

**Comment by samsonasik** [comment]
@TomasVotruba rebased.

**Comment by TomasVotruba** [comment]
Thanks :+1: 

Would adding to parallel if branch help? It seems it's missing by accident there.

```php
     // 2. PHPStan has to know about all files too 
     $this->configurePHPStanNodeScopeResolver($files); 
```

**Comment by samsonasik** [comment]
Call `configurePHPStanNodeScopeResolver` still not works even call in paralell processor.

**Pull request closed without merging**